src/Modal/Modal.tsx

```diff
@@ -0,0 +1,212 @@
+}
+
+const ModalIn = keyframes`
+  0% {
+    opacity: 0;
+    transform: translate(0, -25%);
+  }
+  100% {
+    opacity: 1;
+    transform: translateY(0);
+  }
+`;
+
+const ModalOut = keyframes`
+  0% {
+    opacity: 1;
+    transform: translateY(0);
+  }
+  100% {
+    opacity: 0;
+    transform: translate(0, -25%);
+  }
+`;
+
+const ModalContent = styled(Box)`
+  opacity: 0;
+  transform: translate(0, -25%);
+  overflow: hidden;
+  &[data-state='entering'],
+  &[data-state='entered'] {
+    animation-fill-mode: forwards;
+    animation-name: ${ModalIn};
+    animation-duration: ${ANIMATION_DURATION}ms;
+  }
+  &[data-state='exiting'] {
+    animation-fill-mode: forwards;
+    animation-name: ${ModalOut};
+    animation-duration: ${ANIMATION_DURATION}ms;
+  }
+  &.entered {
+    opacity: 1;
+    transform: translate(0, 0);
+  }
+`;
+
+class Modal extends React.Component<ModalProps, ModalState> {
+  static displayName = 'Modal';
+
+  static defaultProps = {
+    className: undefined,
+    style: undefined,
+    enableFocusTrap: false,
+    disableOverlayClick: false,
+    isOpen: false,
+  };
+
+  constructor(props: ModalProps) {
+    super(props);
+    this.state = {
+      isOpen: props.isOpen,
+    };
+
+    this.handleCloseSideSheet = this.handleCloseSideSheet.bind(this);
+    this.handleOverlayClick = this.handleOverlayClick.bind(this);
+    this.handleKeyDown = this.handleKeyDown.bind(this);
+  }
+
+  static getDerivedStateFromProps(props: ModalProps) {
+    return {
+      isOpen: props.isOpen,
+    };
+  }
+
+  componentDidUpdate(prev: ModalProps) {
+    const { isOpen } = this.props;
+
+    if (prev.isOpen !== isOpen) {
+      if (isOpen) {
+        document.body.classList.add('noscroll');
+      } else {
+        document.body.classList.remove('noscroll');
+      }
+    }
+  }
+
+  handleKeyDown(event: React.KeyboardEvent) {
+    if (event.key === 'Escape') {
+      this.handleCloseSideSheet();
+    }
+  }
+
+  handleCloseSideSheet() {
+    const { onClose } = this.props;
+
+    if (onClose) {
+      onClose();
+    }
+  }
+
+  handleOverlayClick(e: React.MouseEvent<HTMLDivElement>) {
+    const { disableOverlayClick } = this.props;
+
+    // Prevent clicking to exit inside the dialog
+    if (e.target !== e.currentTarget) {
+      return;
+    }
+
+    if (!disableOverlayClick) {
+      this.handleCloseSideSheet();
+    }
+  }
+
+  renderInnerContent = (state: TransitionStatus) => {
+    const { labelledById, hideCloseButton, children } = this.props;
+    const { isOpen } = this.state;
+
+    return (
+      <ModalOverlay className={clsx(isOpen && 'entered')} data-state={state} onClick={this.handleOverlayClick}>
+        <ModalContent
+          className={clsx(isOpen && 'entered')}
+          display="flex"
+          flexDirection="column"
+          backgroundColor="white"
+          boxShadow={2}
+          width="100%"
+          maxWidth="500px"
+          height="100%"
+          maxHeight="calc(100% - 24vmin)"
```

Comment:
Some of these style props may be better off if lifted into the SC rather than having them in the JSX